src/plugins/formulas/formulas.js

```diff
@@ -126,318 +162,427 @@ export class Formulas extends BasePlugin {
    *
+   * @param {Array} sourceData Array of arrays or array of objects containing data.
+   * @param {boolean} initialLoad Flag that determines whether the data has been loaded during the initialization.
+   * @param {string} source Source of the call.
    * @private
-   * @param {Array} cells An array of recalculated/changed cells.
    */
-  onSheetAfterRecalculate(cells) {
-    if (this._skipRendering) {
-      this._skipRendering = false;
-
+  onAfterLoadData(sourceData, initialLoad, source) {
+    if ((source || '').includes(toUpperCaseFirst(PLUGIN_KEY))) {
       return;
     }
-    const hot = this.hot;
 
-    arrayEach(cells, (cellValue) => {
-      if (cellValue instanceof CellValue) {
-        const { row, column } = cellValue;
+    const sheetName = this.hot.getSettings()[PLUGIN_KEY].sheetName;
 
-        hot.validateCell(hot.getDataAtCell(row, column), hot.getCellMeta(row, column), () => {});
+    if (isDefined(sheetName) && this.engine.doesSheetExist(sheetName)) {
+      this.sheetName = sheetName;
+
+    } else {
+      this.sheetName = this.engine.addSheet(sheetName);
+    }
+
+    if (!this.#hotWasInitializedWithEmptyData) {
+      const sourceDataArray = this.hot.getSourceDataArray();
+
+      if (this.engine.isItPossibleToReplaceSheetContent(this.sheetName, sourceDataArray)) {
+        this.#internalOperationPending = true;
+
+        this.engine.setSheetContent(this.sheetName, this.hot.getSourceDataArray());
+
+        this.#internalOperationPending = false;
       }
-    });
-    hot.render();
+
+    } else {
+      this.switchSheet(this.sheetName);
+    }
   }
 
   /**
-   * On modify row data listener. It overwrites raw values into calculated ones and force upper case all formula expressions.
+   * `modifyData` hook callback.
    *
    * @private
-   * @param {number} row Row index.
-   * @param {number} column Column index.
-   * @param {object} valueHolder Value holder as an object to change value by reference.
-   * @param {string} ioMode IO operation (`get` or `set`).
+   * @param {number} row Physical row height.
+   * @param {number} column Physical column index.
+   * @param {object} valueHolder Object which contains original value which can be modified by overwriting `.value`
+   *   property.
+   * @param {string} ioMode String which indicates for what operation hook is fired (`get` or `set`).
    */
   onModifyData(row, column, valueHolder, ioMode) {
-    if (ioMode === 'get' && this.hasComputedCellValue(row, column)) {
-      valueHolder.value = this.getCellValue(row, column);
+    if (!this.enabled || this.#internalOperationPending) {
+      // TODO check if this line is actually ever reached
+      return;
+    }
 
-    } else if (ioMode === 'set' && isFormulaExpression(valueHolder.value)) {
-      valueHolder.value = toUpperCaseFormula(valueHolder.value);
+    const address = {
+      row: this.hot.toVisualRow(row),
+      col: column,
+      sheet: this.engine.getSheetId(this.sheetName)
```

Comment:
We have a getter for that
```suggestion
      sheet: this.sheetId
```

Reply:
https://github.com/handsontable/handsontable/pull/7648/commits/783607d9d8a39bd6ec41d94dfc6e28a451cb4a9a